src/herder/TxSetUtils.cpp

```diff
@@ -0,0 +1,281 @@
+    bool
+    operator()(TxSetFrame::AccountTransactionQueue const* tx1,
+               TxSetFrame::AccountTransactionQueue const* tx2) const
+    {
+        if (tx1 == nullptr || tx1->empty())
+        {
+            return tx2 ? !tx2->empty() : false;
+        }
+        if (tx2 == nullptr || tx2->empty())
+        {
+            return false;
+        }
+
+        auto& top1 = tx1->front();
+        auto& top2 = tx2->front();
+
+        auto cmp3 = feeRate3WayCompare(top1, top2);
+
+        if (cmp3 != 0)
+        {
+            return cmp3 < 0;
+        }
+        // use hash of transaction as a tie breaker
+        return lessThanXored(top1->getFullHash(), top2->getFullHash(), mSeed);
+    }
+};
+
+TxSetFramePtr
+surgePricingFilter(Application& app, TxSetFrameConstPtr txSetPtr)
+{
+    ZoneScoped;
+    LedgerTxn ltx(app.getLedgerTxnRoot());
+    auto header = ltx.loadHeader();
+
+    bool maxIsOps = protocolVersionStartsFrom(header.current().ledgerVersion,
+                                              ProtocolVersion::V_11);
+
+    size_t opsLeft = app.getLedgerManager().getLastMaxTxSetSizeOps();
+
+    auto curSizeOps =
+        maxIsOps ? txSetPtr->sizeOp() : (txSetPtr->sizeTx() * MAX_OPS_PER_TX);
+    if (curSizeOps > opsLeft)
+    {
+        CLOG_WARNING(Herder, "surge pricing in effect! {} > {}", curSizeOps,
+                     opsLeft);
+
+        auto actTxQueueMap = buildAccountTxQueues(*txSetPtr);
+
+        std::priority_queue<TxSetFrame::AccountTransactionQueue*,
+                            std::vector<TxSetFrame::AccountTransactionQueue*>,
+                            SurgeCompare>
+            surgeQueue;
+
+        for (auto& am : actTxQueueMap)
+        {
+            surgeQueue.push(&am.second);
+        }
+
+        TxSetFrame::Transactions updatedSet;
+        updatedSet.reserve(txSetPtr->sizeTx());
+        while (opsLeft > 0 && !surgeQueue.empty())
+        {
+            auto cur = surgeQueue.top();
+            surgeQueue.pop();
+            // inspect the top candidate queue
+            auto& curTopTx = cur->front();
+            size_t opsCount =
+                maxIsOps ? curTopTx->getNumOperations() : MAX_OPS_PER_TX;
+            if (opsCount <= opsLeft)
+            {
+                // pop from this one
+                updatedSet.emplace_back(curTopTx);
+                cur->pop_front();
+                opsLeft -= opsCount;
+                // if there are more transactions, put it back
+                if (!cur->empty())
+                {
+                    surgeQueue.push(cur);
+                }
+            }
+            else
+            {
+                // drop this transaction -> we need to drop the others
+                cur->clear();
+            }
+        }
+
+        return std::make_shared<TxSetFrame>(txSetPtr->previousLedgerHash(),
+                                            updatedSet, false);
+    }
+    return std::make_shared<TxSetFrame>(*txSetPtr);
```

Comment:
We don't need to allocate a new tx set in this case, do we? If we never mutate the tx set, we could just always use the const ptr, so that we don't need to do the cast (which, I suppose, was the reason to allocate a copy here).